### PR TITLE
Issue/1183 compiler service

### PR DESCRIPTION
**Pull request by wouterdb**
Major rework of server compiles

introduces some innovations
1. typing of data objects with the instance types (instead of Field)
2. data v2.py
3. fix for protocol to allow different operations on one url to be in different slices 
4. typing of options

### CODE REVIEW

File: src/inmanta/data.py [hunk 1]
```diff
@@ -304,7 +304,7 @@ def _get_column_names_and_values(self) -> Tuple[List[str], List[str]]:
 
         return (column_names, values)
 
-    async def insert(self, connection=None):
+    async def insert(self, connection=None) -> None:
```
Comment (bartv) [inline src/inmanta/data.py]: also type the arguments

File: src/inmanta/data.py [hunk 2]
```diff
@@ -1130,34 +1130,62 @@ class Report(BaseDocument):
...
+    errstream: str = Field(field_type=str, default="")
+    outstream: str = Field(field_type=str, default="")
+    returncode: int = Field(field_type=int)
+    compile: uuid.UUID = Field(field_type=uuid.UUID, required=True)
+
+    async def update_streams(self, out="", err="") -> None:
+        if not out and not err:
+            return
+        await self._execute_query(
+            f"UPDATE {self.table_name()} SET outstream = outstream || $1, errstream = errstream || $2 WHERE id = $3",
```
Comment (bartv) [inline src/inmanta/data.py]: Should we check that out and err end with a new line? and otherwise add it?

File: src/inmanta/data.py [hunk 3]
```diff
@@ -1130,34 +1130,62 @@ class Report(BaseDocument):
...
+        :param handled: were all registered handlers executed?
+        :param version: version exported by this compile
     """
 
-    id = Field(field_type=uuid.UUID, required=True, part_of_primary_key=True)
-    environment = Field(field_type=uuid.UUID, required=True)
-    started = Field(field_type=datetime.datetime)
-    completed = Field(field_type=datetime.datetime)
+    id: uuid.UUID = Field(field_type=uuid.UUID, required=True, part_of_primary_key=True)
+    remote_id: uuid.UUID = Field(field_type=uuid.UUID)
```
Comment (bartv) [inline src/inmanta/data.py]: if its not required should we make it Optional[uuid.UUID] and assign None? And same for other fields

File: src/inmanta/data.py [hunk 4]
```diff
@@ -1130,34 +1130,62 @@ class Report(BaseDocument):
...
+    metadata: dict = Field(field_type=dict, default={})
+    environment_variables: dict = Field(field_type=dict)
+
+    success: bool = Field(field_type=bool)
+    handled: bool = Field(field_type=bool, default=False)
+    version: int = Field(field_type=int)
 
     @classmethod
-    async def get_reports(cls, environment_id, limit=None, start=None, end=None):
+    async def get_reports(cls, environment_id, limit=None, start=None, end=None) -> List[JsonType]:
```
Comment (bartv) [inline src/inmanta/data.py]: type the arguments?

File: src/inmanta/server/bootloader.py [hunk 5]
```diff
@@ -69,14 +70,8 @@ def __init__(self, agent_no_log: bool = False) -> None:
...
 
-    def get_agent_manager_slice(self) -> AgentManager:
-        return AgentManager()
-
-    def get_database_slice(self) -> DatabaseSlice:
-        return DatabaseSlice()
-
     def get_bootstrap_slices(self) -> List[ServerSlice]:
-        return [self.get_server_slice(), self.get_agent_manager_slice(), self.get_database_slice()]
+        return [self.get_server_slice(), AgentManager(), DatabaseSlice(), CompilerService()]
```
Comment (bartv) [inline src/inmanta/server/bootloader.py]: Put server slice here as well for consistency

File: src/inmanta/server/compilerservice.py [hunk 6]
```diff
@@ -0,0 +1,434 @@
...
+
+from inmanta.server import config as opt
+from inmanta.util import ensure_directory_exist
+
+LOGGER = logging.getLogger(__name__)
+
+
+class CompileStateListener(object):
+    @abc.abstractmethod
+    async def compile_done(self, compile: data.Compile):
```
Comment (bartv) [inline src/inmanta/server/compilerservice.py]: no return type

File: src/inmanta/server/compilerservice.py [hunk 7]
```diff
@@ -0,0 +1,434 @@
...
+            sub_process = await asyncio.create_subprocess_exec(
+                cmd[0], *cmd[1:], stdout=subprocess.PIPE, stderr=subprocess.PIPE, cwd=cwd, env=env
+            )
+
+            returncode = await self.drain(sub_process)
+
+            return await self._end_stage(returncode)
+        except Exception as e:
+            await self._error("".join(traceback.format_exception(type(e), e, e.__traceback__)))
+            return await self._end_stage(-1)
```
Comment (bartv) [inline src/inmanta/server/compilerservice.py]: Make -1 an internal constant so its clear why we use it

File: src/inmanta/server/compilerservice.py [hunk 8]
```diff
@@ -0,0 +1,434 @@
...
+
+        except Exception:
+            LOGGER.exception("An error occured while recompiling")
+
+        finally:
+            return success
+
+
+class CompilerService(ServerSlice):
+
```
Comment (bartv) [inline src/inmanta/server/compilerservice.py]: Add docs that explain queue/deque and how tasks are persisted

File: src/inmanta/data.py (hunk 2 above)
Comment (wouterdb) [reply src/inmanta/data.py]: there is guarantee that chunks read from the stream end on a newline.
This way, we faithfully store the data as it is reported, without adding (or removing) newlines.

File: src/inmanta/data.py [hunk 9]
```diff
@@ -803,7 +803,7 @@ class Environment(BaseDocument):
         AUTOSTART_AGENT_DEPLOY_SPLAY_TIME: AUTOSTART_SPLAY,
     }  # name new_option -> name deprecated_option
 
-    async def get(self, key):
+    async def get(self, key: str):
```
Comment (bartv) [inline src/inmanta/data.py]: no return type

File: src/inmanta/data.py [hunk 10]
```diff
@@ -1130,34 +1130,64 @@ class Report(BaseDocument):
...
+    started: datetime.datetime = Field(field_type=datetime.datetime, required=True)
+    completed: Optional[datetime.datetime] = Field(field_type=datetime.datetime)
+    command: str = Field(field_type=str, required=True)
+    name: str = Field(field_type=str, required=True)
+    errstream: str = Field(field_type=str, default="")
+    outstream: str = Field(field_type=str, default="")
+    returncode: Optional[int] = Field(field_type=int)
+    compile: uuid.UUID = Field(field_type=uuid.UUID, required=True)
+
+    async def update_streams(self, out="", err="") -> None:
```
Comment (bartv) [inline src/inmanta/data.py]: no argument types

File: src/inmanta/data.py [hunk 11]
```diff
@@ -1884,25 +1957,25 @@ def new(cls, environment, resource_version_id, **kwargs):
 
         return should_purge
 
-    async def insert(self, connection=None):
+    async def insert(self, connection=None) -> None:
```
Comment (bartv) [inline src/inmanta/data.py]: no argument types

File: src/inmanta/data.py [hunk 12]
```diff
@@ -1884,25 +1957,25 @@ def new(cls, environment, resource_version_id, **kwargs):
...
         self.make_hash()
         await super(Resource, self).insert(connection=connection)
 
     @classmethod
-    async def insert_many(cls, documents):
+    async def insert_many(cls, documents) -> None:
```
Comment (bartv) [inline src/inmanta/data.py]: no argument types

File: src/inmanta/data.py [hunk 13]
```diff
@@ -1884,25 +1957,25 @@ def new(cls, environment, resource_version_id, **kwargs):
...
         for doc in documents:
             doc.make_hash()
         await super(Resource, cls).insert_many(documents)
 
-    async def update(self, **kwargs):
+    async def update(self, **kwargs) -> None:
```
Comment (bartv) [inline src/inmanta/data.py]: no argument types

File: src/inmanta/data.py [hunk 14]
```diff
@@ -1884,25 +1957,25 @@ def new(cls, environment, resource_version_id, **kwargs):
...
         self.make_hash()
         await super(Resource, self).update(**kwargs)
 
-    async def update_fields(self, **kwargs):
+    async def update_fields(self, **kwargs) -> None:
```
Comment (bartv) [inline src/inmanta/data.py]: no argument types

File: src/inmanta/server/server.py [hunk 15]
```diff
@@ -1971,222 +1973,17 @@ def safe_get(input, key, default):
         if "message" not in metadata:
             metadata["message"] = "Recompile trigger through API call"
 
-        await self._async_recompile(env, update, metadata=metadata)
+        warnings = await self._async_recompile(env, update, metadata=metadata)
 
-        return 200
+        return attach_warnings(200, None, warnings)
 
     async def _async_recompile(self, env: data.Environment, update_repo: bool, metadata: JsonType = {}) -> None:
```
Comment (arnaudsjs) [inline src/inmanta/server/server.py]: _async_recompile always returns None, while many methods use its return type.

File: src/inmanta/server/compilerservice.py [hunk 16]
```diff
@@ -0,0 +1,462 @@
...
+        for u in unhandled:
+            self.add_background_task(self._notify_listeners(u))
+
+    @protocol.handle(methods.is_compiling, environment_id="id")
+    async def is_compiling(self, environment_id: uuid.UUID) -> Apireturn:
+        if environment_id in self._recompiles and not self._recompiles[environment_id].done():
+            return 200
+        return 204
+
+    async def _run(self, compile: data.Compile):
```
Comment (bartv) [inline src/inmanta/server/compilerservice.py]: no return type

**Comment by bartv** [comment]
Does this PR also fix #184 ?

**Comment by wouterdb** [comment]
@bart: no it doesn't fix #184, as there is no API endpoint to change the branch

File: src/inmanta/data.py [hunk 17]
```diff
@@ -1130,34 +1130,65 @@ class Report(BaseDocument):
...
+    environment_variables: dict = Field(field_type=dict)
+
+    success: Optional[bool] = Field(field_type=bool)
+    handled: bool = Field(field_type=bool, default=False)
+    version: Optional[int] = Field(field_type=int)
 
     @classmethod
-    async def get_reports(cls, environment_id, limit=None, start=None, end=None):
+    async def get_reports(
+        cls, environment_id: uuid.UUID, limit: int = None, start: datetime.datetime = None, end: datetime.datetime = None
```
Comment (arnaudsjs) [inline src/inmanta/data.py]: Attributes limit, start and end are optional

File: src/inmanta/server/compilerservice.py [hunk 18]
```diff
@@ -0,0 +1,464 @@
...
+
+    def _add_to_tail(self, part: str) -> None:
+        if self.tail_stdout is not None:
+            self.tail_stdout += part
+            self.tail_stdout = self.tail_stdout[-1024:]
+
+    async def drain_out(self, stream: asyncio.StreamReader) -> None:
+        assert self.stage is not None
+        while not stream.at_eof():
+            part = (await stream.read(1)).decode()
```
Comment (arnaudsjs) [inline src/inmanta/server/compilerservice.py]: Stream is read one byte at a time

File: src/inmanta/server/compilerservice.py [hunk 19]
```diff
@@ -0,0 +1,464 @@
...
+            self._add_to_tail(part)
+            await self.stage.update_streams(out=part)
+
+    async def drain_err(self, stream: asyncio.StreamReader) -> None:
+        assert self.stage is not None
+        while not stream.at_eof():
+            part = (await stream.read(1)).decode()
```
Comment (arnaudsjs) [inline src/inmanta/server/compilerservice.py]: Stream is read one byte at a time

**Pull request merged**